Fix borrowing issues

SabrinaJewson opened the pull request:
Previously, code like this would fail to compile:

```rust
#[context("context")]
async fn borrows(val: &mut u32) -> anyhow::Result<&u32> {
    Ok(&*val)
}
```

With this error message:

```
error[E0373]: async block may outlive the current function, but it borrows `val`, which is owned by the current function
 --> src/lib.rs:4:1
  |
4 | #[context("context")]
  | ^^^^^^^^^^^^^^^^^^^^^ may outlive borrowed value `val`
5 | async fn borrows(val: &mut u32) -> anyhow::Result<&u32> {
6 |     Ok(&*val)
  |          --- `val` is borrowed here
  |
note: async block is returned here
 --> src/lib.rs:4:1
  |
4 | #[context("context")]
  | ^^^^^^^^^^^^^^^^^^^^^
  = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)
help: to force the async block to take ownership of `val` (and any other referenced variables), use the `move` keyword
  |
4 | move #[context("context")]
  |
```

This can be fixed by using an `async move {}` block instead of an `async {}` block.

Commit:
Use `async move` block to fix borrowing issues

andrewhickman commented:
Thanks for the PR!

Looks like the sync version has the same issue. There's a bit of a tradeoff between supporting code like

```rust
#[context("failed to parse config at `{}`", path.as_ref().display())]
pub fn parse_config(path: impl AsRef<Path>) -> anyhow::Result<u32> {
    let text = read_to_string(path.as_ref())?;
    Ok(text.parse()?)
}
```

which fails when we use a move closure/async move block, and your example.

Do you think a way to opt into the `move` semantics would be reasonable? e.g.

```rust
#[context(move, "context")]
async fn borrows(val: &mut u32) -> anyhow::Result<&u32> {
```

SabrinaJewson commented:
Oh, you're right - I didn't consider that case. `#[context(move)]` seems like a better solution. I can implement it tomorrow in this PR, but feel free to do it now if you want.

Commit:
Add support for `move` option

Commit:
Force the closure to be `FnOnce`

SabrinaJewson commented:
I have updated the PR to add support for the `move` option. Additionally, I implemented a commit that should reduce incorrect borrowck errors in synchronous functions by forcing the closure to be `FnOnce`.

andrewhickman commented on src/lib.rs:
```diff
@@ -55,32 +58,64 @@ use quote::quote;
+        quote! {
+            // Moving a non-`Copy` value into the closure tells borrowck to always treat the closure
+            // as a `FnOnce`, preventing some borrowing errors.
+            let force_fn_once = ::core::iter::empty::<()>();
```
I think we need to use an ugly name like `__force_fn_once` or it might conflict with variable in the function body

I did a quick search and it sounds like we could instead do `Ident::new("force_fn_once", Span::mixed_site())` (https://github.com/rust-lang/rust/pull/64690) which would be a bit cleaner

SabrinaJewson replied:
I don't think this is _technically_ necessary because the variable is moved out of before any user code is run, but I suppose it's good practice to be hygienic :)

andrewhickman commented on src/lib.rs:
```diff
@@ -55,32 +58,64 @@ use quote::quote;
+            syn::ReturnType::Type(_, return_ty) => return_ty,
+        };
+        quote! {
+            let result: #return_ty = async #move_token { #body }.await;
```
The comment below applies here to `result` as well

andrewhickman reviewed (commented):
Looks great, thanks! The trick with forcing the closure to be `FnOnce` is quite cool. I just had a couple comments about variable hygiene, then it should be ready to merge

Commit:
Use mixed-site hygiene on all locals

andrewhickman commented:
Thanks for the PR! I'll release version 0.2.0 with these changes
